lucille-core/src/main/java/com/kmwllc/lucille/stage/QueryOpensearch.java

```diff
@@ -0,0 +1,177 @@
+ *
+ *
+ * TODO: Is that needed?
+ * requiredParamNames (list of String, optional): A list of field names that you require to be found on every document. This list should...
+ *   1.) Contain all parameters you defined in your search template that do <b>not</b> have default values
+ *   2.) Be found on every Document you intend to process.
+ * If any of the required parameters are missing, Lucille will throw an Exception to prevent unexpected query results from being applied to Documents.
+ *
+ * optionalParamNames (list of String, optional): A list of field names that you do not require to be found on every document, but want
+ * to include in your query if they are found. This list should...
+ *   1.) contain all parameters you defined in your search template that have default values
+ *   2.) The same field names should be found on Documents (when present / defined).
+ * If any of the optional parameters are missing, they will not be used in the search, and the default value will be used instead.
+ *
+ * opensearchResponsePath (String, Optional): A path to a field in the Opensearch response whose value you want to place on a Lucille Document.
+ * Use JsonPointer notation. An IllegalArgumentException will be thrown if the path has invalid formatting. Defaults to using the entire
+ * response.
+ *
+ * destinationField (String, Optional): The name of the field you'll write the response value to in a Lucille Document. Defaults to "response".
+ */
+public class QueryOpensearch extends Stage {
+
+  private final URI searchURI;
+  private final URI templateURI;
+
+  private final String templateName;
+  private final String searchTemplate;
+  private final List<String> paramNames;
+  private final JsonPointer opensearchResponsePath;
+  private final String destinationField;
+
+  private final HttpClient httpClient;
+
+  public QueryOpensearch(Config config) {
+    super(config, new StageSpec()
+        .withRequiredParents("opensearch")
+        .withRequiredProperties("templateName", "searchTemplate", "paramNames")
+        .withOptionalProperties("opensearchResponsePath", "destinationField"));
+
+    this.searchURI = getSearchURI(config);
+    this.templateURI = getTemplateURI(config);
+
+    this.templateName = config.getString("templateName");
+    this.searchTemplate = config.getString("searchTemplate");
+    this.paramNames = config.getStringList("paramNames");
+
+    if (config.hasPath("opensearchResponsePath")) {
+      this.opensearchResponsePath = JsonPointer.compile(config.getString("opensearchResponsePath"));
+    } else {
+      this.opensearchResponsePath = JsonPointer.empty();
+    }
+
+    this.destinationField = ConfigUtils.getOrDefault(config, "destinationField", "response");
+
+    httpClient = HttpClient.newHttpClient();
+  }
+
+  @Override
+  public void start() throws StageException {
+    // Register the search template with Opensearch.
+    HttpRequest request = HttpRequest.newBuilder()
```

Comment:
for the first pass at this stage, I don't think we should provide the ability to save the template in OpenSearch; instead, we can send the entire user-provided query with the params on every request. If the user has _already_ saved the template in Opensearch, they can provide a simplified query in the stage config that references the saved template by its ID. If the user hasn't saved the template, they can provide the full body of the query...

Reply:
I just mean we can take the query the user provides and fill in the params section of it. They have the choice of providing a full template (which would then get sent to opensearch on every request, but never saved) or they can include json that references an already saved template. Can discuss further...